### PR TITLE
feature save and load

### PR DESCRIPTION
**Pull request by bxy4543**
feature sealer save and load: save a CloudImage to tar.gz, and load command can load it;

### CODE REVIEW

File: image/default_image_file.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	var image v1.Image
+	if err := utils.UnmarshalYamlFile(imageTempFile, &image); err != nil {
+		return fmt.Errorf("%s.yaml parsing failed, %s", imageTempFile, err)
+	}
+	tempfile, err := os.Open(imageTempFile)
+	if err != nil {
+		return fmt.Errorf("failed to create image_metadata.yaml,%v", err)
+	}
+	defer utils.CleanFile(tempfile)
+	imageYamlFile := filepath.Join(common.DefaultImageMetaRootDir, image.Spec.ID+common.YamlSuffix)
```
Comment (justadogistaken) [inline image/default_image_file.go:114]: there is a func syncImageLocal to save image

File: image/default_image_file.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	file, err := os.Create(imageYamlFile)
+	if err != nil {
+		return fmt.Errorf("failed to write image metadata,%v", err)
+	}
+	defer file.Close()
+	_, err = io.Copy(file, tempfile)
+	if err != nil {
+		return fmt.Errorf("failed to write image metadata file,%v", err)
+	}
+	for _, layer := range image.Spec.Layers {
```
Comment (justadogistaken) [inline image/default_image_file.go:129]: use layerStore register layer

File: sealer/cmd/login.go [hunk 3]
```diff
@@ -52,4 +53,12 @@ func init() {
 	rootCmd.AddCommand(loginCmd)
 	loginCmd.Flags().StringVarP(&loginConfig.RegistryUsername, "username", "u", "", "user name for login registry")
 	loginCmd.Flags().StringVarP(&loginConfig.RegistryPasswd, "passwd", "p", "", "password for login registry")
+	if err := loginCmd.MarkFlagRequired("username"); err != nil {
+		logger.Error("failed to init flag: %v", err)
+		os.Exit(1)
+	}
+	if err := loginCmd.MarkFlagRequired("passwd"); err != nil {
+		logger.Error("failed to init flag: %v")
```
Comment (fanux) [inline sealer/cmd/login.go:61]: %v",err)

File: image/default_image_file.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	}
+	pathToCompress = append(pathToCompress, imageMetadataTempFile)
+	_, err = compress.Compress(file, pathToCompress...)
+	return err
+}
+
+func (d DefaultImageFileService) load(imageSrc string) error {
+	src, err := os.Open(imageSrc)
+	if err != nil {
+		return fmt.Errorf("failed to open %s,%v", imageSrc, err)
```
Comment (justadogistaken) [inline image/default_image_file.go:98]: please unify the error format.  "failed to open %s, err: %v"  . Be serious about the format

File: image/default_image_file.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+	}
+	defer src.Close()
+	if err := compress.Decompress(src, common.DefaultLayerDir); err != nil {
+		return err
+	}
+	imageTempFile := filepath.Join(common.DefaultLayerDir, common.DefaultImageMetadataFileName)
+	var image v1.Image
+	if err := utils.UnmarshalYamlFile(imageTempFile, &image); err != nil {
+		return fmt.Errorf("%s.yaml parsing failed, %s", imageTempFile, err)
```
Comment (justadogistaken) [inline image/default_image_file.go:107]: format, check all occurrences

File: image/default_image_file.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+	if imageTar == "" {
+		imageTar = ima.Spec.ID + common.TarGzSuffix
+	}
+	if _, err := os.Stat(imageTar); err == nil {
+		return fmt.Errorf("file %s exist", imageTar)
+	}
+	err = os.MkdirAll(filepath.Dir(imageTar), common.FileMode0644)
+	if err != nil {
+		return fmt.Errorf("failed to create %s,err:%v", imageTar, err)
+	}
```
Comment (justadogistaken) [inline image/default_image_file.go:55]: Maybe the above jobs could be put into Save.

File: image/default_image_file.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+	file, err := os.Create(imageTar)
+	if err != nil {
+		return fmt.Errorf("failed to create %s,err:%v", imageTar, err)
+	}
+	defer file.Close()
+
+	var pathToCompress []string
```
Comment (justadogistaken) [inline image/default_image_file.go:62]: pathsToCompress

File: image/default_image_file.go [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+	if _, err := os.Stat(imageTar); err == nil {
+		return fmt.Errorf("file %s exist", imageTar)
+	}
+	err = os.MkdirAll(filepath.Dir(imageTar), common.FileMode0644)
+	if err != nil {
+		return fmt.Errorf("failed to create %s,err:%v", imageTar, err)
+	}
+	file, err := os.Create(imageTar)
+	if err != nil {
+		return fmt.Errorf("failed to create %s,err:%v", imageTar, err)
```
Comment (justadogistaken) [inline image/default_image_file.go:58]: better leave a space following "," ":"

File: image/default_image_file.go [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+	if err != nil {
+		return fmt.Errorf("failed to create %s:%v", tempDir, err)
+	}
+	defer utils.CleanDir(tempDir)
+	imageMetadataTempFile := filepath.Join(tempDir, common.DefaultImageMetadataFileName)
+	tempfile, err := os.Create(imageMetadataTempFile)
+	if err != nil {
+		return fmt.Errorf("failed to create image_metadata.yaml,err:%v", err)
+	}
+	imageMetadataFile, err := os.Open(filepath.Join(common.DefaultImageMetaRootDir, ima.Spec.ID+common.YamlSuffix))
```
Comment (justadogistaken) [inline image/default_image_file.go:81]: do not have to open it, just copy to temp file

File: image/default_image_file.go [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+	if err != nil {
+		return err
+	}
+	ima, err := imageutils.GetImage(named.Raw())
+	if err != nil {
+		return err
+	}
+	if imageTar == "" {
+		imageTar = ima.Spec.ID + common.TarGzSuffix
+	}
```
Comment (justadogistaken) [inline image/default_image_file.go:56]: you checked {imageTar == "" } in Save already.

File: image/default_image_file.go [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+		return fmt.Errorf("failed to create %s, err: %v", imageTar, err)
+	}
+	defer file.Close()
+
+	var pathsToCompress []string
+	for _, layer := range ima.Spec.Layers {
+		if layer.Hash == "" {
+			continue
+		}
+		layerID := store.LayerID(digest.NewDigestFromEncoded(digest.SHA256, layer.Hash.String()))
```
Comment (justadogistaken) [inline image/default_image_file.go:68]: layer.Hash is digest.Digest at present, you may need to rebase upstream main

File: image/default_image_file.go [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+	}
+	defer utils.CleanDir(tempDir)
+	imageMetadataTempFile := filepath.Join(tempDir, common.DefaultImageMetadataFileName)
+	imageMetadataYaml := filepath.Join(common.DefaultImageMetaRootDir, ima.Spec.ID+common.YamlSuffix)
+	imageMetadata, err := ioutil.ReadFile(imageMetadataYaml)
+	if err != nil {
+		return fmt.Errorf("failed to read %s, err: %v", imageMetadataYaml, err)
+	}
+	if err := ioutil.WriteFile(imageMetadataTempFile, imageMetadata, common.FileMode0644); err != nil {
+		return fmt.Errorf("failed to write temp file %s, err :%v ", imageMetadataTempFile, err)
```
Comment (justadogistaken) [inline image/default_image_file.go:84]: err:

File: image/default_image_file.go [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+	if err := utils.UnmarshalYamlFile(imageTempFile, &image); err != nil {
+		return fmt.Errorf("failed to parsing %s.yaml, err: %v", imageTempFile, err)
+	}
+	defer os.Remove(imageTempFile)
+
+	for _, layer := range image.Spec.Layers {
+		if layer.Hash == "" {
+			continue
+		}
+		layerID := store.LayerID(digest.NewDigestFromEncoded(digest.SHA256, layer.Hash.String()))
```
Comment (justadogistaken) [inline image/default_image_file.go:111]: rebase upstream/main, layer.Hash is digest.Digest

File: image/default_image_file.go [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+	if err != nil {
+		return fmt.Errorf("failed to open %s, err : %v", imageSrc, err)
+	}
+	defer src.Close()
+	if err := compress.Decompress(src, common.DefaultLayerDir); err != nil {
+		return err
+	}
+	imageTempFile := filepath.Join(common.DefaultLayerDir, common.DefaultImageMetadataFileName)
+	var image v1.Image
+	if err := utils.UnmarshalYamlFile(imageTempFile, &image); err != nil {
```
Comment (justadogistaken) [inline image/default_image_file.go:96]: err has been initiated above

File: sealer/cmd/load.go [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+	Short:   "load image",
+	Long:    `Load an image from a tar archive`,
+	Example: `sealer load -i kubernetes.tar.gz`,
+	Args:    cobra.ExactArgs(0),
+	Run: func(cmd *cobra.Command, args []string) {
+		if err := image.NewImageFileService().Load(imageSrc); err != nil {
+			logger.Error("failed to load image from %s, err: %v", imageSrc, err)
+			os.Exit(1)
+		}
+		logger.Info("load image from %s successfully", imageSrc)
```
Comment (justadogistaken) [inline sealer/cmd/load.go:25]: need to let user know the final image tag and id.

File: image/default_image_file.go [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+	return d.load(imageSrc)
+}
+
+func (d DefaultImageFileService) Save(imageName string, imageTar string) error {
+	if imageTar == "" {
+		return fmt.Errorf("imagetar cannot be empty")
+	}
+	if utils.IsFileExist(imageTar) {
+		return fmt.Errorf("file %s already exists", imageTar)
+	}
```
Comment (justadogistaken) [inline image/default_image_file.go:34]: I suggest that adding a line following some keyword like "if".

File: image/default_image_file.go [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+		return fmt.Errorf("failed to parsing %s.yaml, err: %v", imageTempFile, err)
+	}
+
+	layerStore, err := store.NewDefaultLayerStore()
+	if err != nil {
+		return fmt.Errorf("failed to get layerstore, err: %v", err)
+	}
+
+	for _, layer := range image.Spec.Layers {
+		if layer.Hash != "" {
```
Comment (justadogistaken) [inline image/default_image_file.go:125]: if layer.Hash == "" {continue}, do it in this way, which can help decrease code nesting

File: image/default_image_file.go [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+	}
+	named, err := reference.ParseToNamed(imageMetadata.Name)
+	if err != nil {
+		return err
+	}
+
+	if err = d.syncImageLocal(image, named); err != nil {
+		return err
+	}
+	logger.Info("load image %s [id: %s] successfully", named.Raw(), imageMetadata.ID)
```
Comment (justadogistaken) [inline image/default_image_file.go:154]: put this log to func Load, and fast return d.syncImageLocal here

**Review by justadogistaken — APPROVED** [review]
LGTM, make sure you have tested before being merged

**Pull request merged**